File: app/src/main/java/info/staticfree/supergenpass/fragment/Preferences.java
package info.staticfree.supergenpass.fragment;

import android.app.LoaderManager;
import android.content.CursorLoader;
import android.content.Intent;
import android.content.Loader;
import android.content.SharedPreferences;
import android.database.Cursor;
import android.os.Bundle;
import android.preference.EditTextPreference;
import android.preference.Preference;
import android.preference.Preference.OnPreferenceChangeListener;
import android.preference.PreferenceFragment;
import android.widget.Toast;

import androidx.annotation.NonNull;

import com.google.zxing.integration.android.IntentIntegrator;
import com.google.zxing.integration.android.IntentResult;

import info.staticfree.supergenpass.db.Domain;
import info.staticfree.supergenpass.R;

public class Preferences extends PreferenceFragment {

    public static final String ACTION_SCAN_SALT =
            "info.staticfree.android.supergenpass.action.SCAN_SALT";
    public static final String ACTION_GENERATE_SALT =
            "info.staticfree.android.supergenpass.action.GENERATE_SALT";

    public static final String ACTION_CLEAR_STORED_DOMAINS =
            "info.staticfree.android.supergenpass.action.CLEAR_STORED_DOMAINS";

    public static final String PREF_PW_TYPE = "pw_type";
    public static final String PREF_PW_LENGTH = "pw_length";
    public static final String PREF_PW_SALT = "pw_salt";
    public static final String PREF_GENERATE_SALT = "generate_salt";
    public static final String PREF_CLIPBOARD = "clipboard";
    public static final String PREF_REMEMBER_DOMAINS = "domain_autocomplete";
    public static final String PREF_DOMAIN_CHECK = "domain_check";
    public static final String PREF_SHOW_GEN_PW = "show_gen_pw";
    public static final String PREF_PW_CLEAR_TIMEOUT = "pw_clear_timeout";
    public static final String PREF_CLEAR_REMEMBERED = "clear_remembered";
    public static final String PREF_SHOW_PIN = "show_pin";
    public static final String PREF_SCAN_SALT = "scan_salt";
    public static final String PREF_PIN_DIGITS = "pw_pin_digits";
    public static final String PREF_VISUAL_HASH = "visual_hash";

    // idea borrowed from
    // http://stackoverflow.com/questions/3206765/number-preferences-in-preference-activity-in
    // -android
    private final OnPreferenceChangeListener integerConformCheck =
            new OnPreferenceChangeListener() {

                @Override
                public boolean onPreferenceChange(Preference preference,
                        Object newValue) {
                    if (!isInteger(newValue)) {
                        Toast.makeText(getActivity().getApplicationContext(),
                                R.string.pref_err_not_number, Toast.LENGTH_LONG).show();
                        return false;
                    }
                    return true;
                }
            };

    private final LoaderManager.LoaderCallbacks<Cursor> mDomainCountLoaderCallbacks =
            new LoaderManager.LoaderCallbacks<Cursor>() {

                @Override
                public Loader<Cursor> onCreateLoader(int id, Bundle args) {
                    return new CursorLoader(getActivity(), Domain.CONTENT_URI, new String[] {},
                            null, null, null);
                }

                @Override
                public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
                    int domainCount = data.getCount();
                    if (isResumed() && !isRemoving()) {
                        Preference clear = findPreference(PREF_CLEAR_REMEMBERED);
                        clear.setEnabled(domainCount > 0);
                        clear.setSummary(getResources()
                                .getQuantityString(R.plurals.pref_autocomplete_count, domainCount,
                                        domainCount));
                    }
                }

                @Override
                public void onLoaderReset(Loader<Cursor> loader) {

                }
            };

    public boolean isInteger(Object newValue) {
        try {
            //noinspection ResultOfMethodCallIgnored
            Integer.parseInt((String) newValue);
        } catch (@NonNull NumberFormatException e) {
            return false;
        }
        return true;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        addPreferencesFromResource(R.xml.preferences);
        findPreference(PREF_PW_CLEAR_TIMEOUT).setOnPreferenceChangeListener(integerConformCheck);
        findPreference(PREF_PW_LENGTH).setOnPreferenceChangeListener(integerConformCheck);

        findPreference(PREF_SCAN_SALT).setOnPreferenceClickListener(mOnPreferenceClickListener);
        findPreference(PREF_CLEAR_REMEMBERED)
                .setOnPreferenceClickListener(mOnPreferenceClickListener);
        findPreference(PREF_GENERATE_SALT).setOnPreferenceClickListener(mOnPreferenceClickListener);
    }

    @Override
    public void onResume() {
        super.onResume();

        getLoaderManager().restartLoader(0, null, mDomainCountLoaderCallbacks);
    }

    public void scanSalt() {
        IntentIntegrator qr = new IntentIntegrator(getActivity());
        qr.addExtra("PROMPT_MESSAGE", getString(R.string.pref_scan_qr_code_to_load_zxing_message));
        qr.addExtra("SAVE_HISTORY", false);
        qr.initiateScan(IntentIntegrator.QR_CODE_TYPES);
    }

    private void setSaltPref(String salt) {
        ((EditTextPreference) findPreference(PREF_PW_SALT)).setText(salt);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode,
            @NonNull Intent data) {
        IntentResult res =
                IntentIntegrator.parseActivityResult(requestCode, resultCode, data);

        if (res != null && res.getContents() != null) {
            String salt = res.getContents();
            setSaltPref(salt);
        }
    }

    public static int getStringAsInteger(@NonNull SharedPreferences prefs, String key,
            int def) {
        String defString = Integer.toString(def);
        int retval;
        try {
            retval = Integer.parseInt(prefs.getString(key, defString));

            // in case the value ever gets corrupt, reset it to the default instead of freaking out
        } catch (@NonNull NumberFormatException e) {
            prefs.edit().putString(key, defString).apply();
            retval = def;
        }
        return retval;
    }

    private final Preference.OnPreferenceClickListener mOnPreferenceClickListener =
            new Preference.OnPreferenceClickListener() {

                @Override
                public boolean onPreferenceClick(Preference preference) {
                    switch (preference.getKey()) {
                        case PREF_SCAN_SALT:
                            scanSalt();
                            return true;
                        case PREF_GENERATE_SALT:
                            new SaltFragment().show(getFragmentManager(), "salt");
                            return true;
                        case PREF_CLEAR_REMEMBERED:
                            getActivity().getContentResolver()
                                    .delete(Domain.CONTENT_URI, null, null);
                            return true;
                    }

                    return false;
                }
            };

}

File: app/src/main/java/info/staticfree/supergenpass/fragment/Preferences.kt
package info.staticfree.supergenpass.fragment

import android.content.Intent
import android.content.SharedPreferences
import android.database.Cursor
import android.os.Bundle
import android.text.InputType
import androidx.loader.app.LoaderManager
import androidx.loader.content.CursorLoader
import androidx.loader.content.Loader
import androidx.preference.EditTextPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.zxing.integration.android.IntentIntegrator
import info.staticfree.supergenpass.R
import info.staticfree.supergenpass.db.Domain

class Preferences : PreferenceFragmentCompat() {
    private val domainCountLoaderCallbacks: LoaderManager.LoaderCallbacks<Cursor> =
        object : LoaderManager.LoaderCallbacks<Cursor> {
            override fun onCreateLoader(id: Int, args: Bundle?): Loader<Cursor> {
                return CursorLoader(
                    requireContext(), Domain.CONTENT_URI, arrayOf(),
                    null, null, null
                )
            }

            override fun onLoadFinished(loader: Loader<Cursor>, data: Cursor) {
                val domainCount = data.count
                if (isResumed && !isRemoving) {
                    val clear: Preference? = findPreference(PREF_CLEAR_REMEMBERED)
                    clear?.apply {
                        isEnabled = domainCount > 0
                        summary = resources
                            .getQuantityString(
                                R.plurals.pref_autocomplete_count, domainCount,
                                domainCount
                            )
                    }
                }
            }

            override fun onLoaderReset(loader: Loader<Cursor>) {}
        }

    override fun onResume() {
        super.onResume()
        LoaderManager.getInstance(this).restartLoader(0, null, domainCountLoaderCallbacks)
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource(R.xml.preferences)
        (findPreference(PREF_PW_CLEAR_TIMEOUT) as EditTextPreference?)?.setOnBindEditTextListener {
            it.inputType = InputType.TYPE_CLASS_NUMBER
        }
        (findPreference(PREF_PW_LENGTH) as EditTextPreference?)?.setOnBindEditTextListener {
            it.inputType = InputType.TYPE_CLASS_NUMBER
        }
        (findPreference(PREF_SCAN_SALT) as Preference?)?.onPreferenceClickListener =
            onPreferenceClickListener
        (findPreference(PREF_CLEAR_REMEMBERED) as Preference?)?.onPreferenceClickListener =
            onPreferenceClickListener
        (findPreference(PREF_GENERATE_SALT) as Preference?)?.onPreferenceClickListener =
            onPreferenceClickListener
    }

    fun scanSalt() {
        val qr = IntentIntegrator(activity)
        qr.addExtra("PROMPT_MESSAGE", getString(R.string.pref_scan_qr_code_to_load_zxing_message))
        qr.addExtra("SAVE_HISTORY", false)
        qr.initiateScan(IntentIntegrator.QR_CODE_TYPES)
    }

    private fun setSaltPref(salt: String) {
        (findPreference(PREF_PW_SALT) as EditTextPreference?)?.text = salt
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        val res = IntentIntegrator.parseActivityResult(requestCode, resultCode, data)
        if (res != null && res.contents != null) {
            val salt = res.contents
            setSaltPref(salt)
        }
    }

    private val onPreferenceClickListener = Preference.OnPreferenceClickListener { preference ->
        when (preference.key) {
            PREF_SCAN_SALT -> {
                scanSalt()
                return@OnPreferenceClickListener true
            }
            PREF_GENERATE_SALT -> {
                SaltFragment().show(childFragmentManager, "salt")
                return@OnPreferenceClickListener true
            }
            PREF_CLEAR_REMEMBERED -> {
                showClearRememberedDomainDialog()
                return@OnPreferenceClickListener true
            }
            else -> false
        }
    }

    private fun showClearRememberedDomainDialog() {
        MaterialAlertDialogBuilder(requireContext())
            .setTitle(R.string.pref_clear_remembered_domains)
            .setMessage(getString(R.string.pref_clear_domain_confirm_message))
            .setPositiveButton(getString(R.string.dialog_button_clear)) { _, _ ->
                requireContext().contentResolver.delete(Domain.CONTENT_URI, null, null)
            }
            .setNegativeButton(android.R.string.cancel) { dialog, _ -> dialog.dismiss() }
            .setCancelable(true)
            .show()
    }

    companion object {
        const val ACTION_SCAN_SALT = "info.staticfree.android.supergenpass.action.SCAN_SALT"
        const val ACTION_GENERATE_SALT = "info.staticfree.android.supergenpass.action.GENERATE_SALT"
        const val ACTION_CLEAR_STORED_DOMAINS =
            "info.staticfree.android.supergenpass.action.CLEAR_STORED_DOMAINS"
        const val PREF_PW_TYPE = "pw_type"
        const val PREF_PW_LENGTH = "pw_length"
        const val PREF_PW_SALT = "pw_salt"
        const val PREF_GENERATE_SALT = "generate_salt"
        const val PREF_CLIPBOARD = "clipboard"
        const val PREF_REMEMBER_DOMAINS = "domain_autocomplete"
        const val PREF_DOMAIN_CHECK = "domain_check"
        const val PREF_SHOW_GEN_PW = "show_gen_pw"
        const val PREF_PW_CLEAR_TIMEOUT = "pw_clear_timeout"
        const val PREF_CLEAR_REMEMBERED = "clear_remembered"
        const val PREF_SHOW_PIN = "show_pin"
        const val PREF_SCAN_SALT = "scan_salt"
        const val PREF_PIN_DIGITS = "pw_pin_digits"
        const val PREF_VISUAL_HASH = "visual_hash"

        @JvmStatic
        fun getStringAsInteger(prefs: SharedPreferences, key: String?, def: Int): Int {
            val defString = def.toString()
            return try {
                prefs.getString(key, defString)!!.toInt()

                // in case the value ever gets corrupt, reset it to the default instead of freaking out
            } catch (e: NumberFormatException) {
                prefs.edit().putString(key, defString).apply()
                def
            }
        }
    }
}